ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,529 @@
+	Index     int
+	Roots     map[*node.Node]uint64 // node index for root nodes
+	NodeCount uint64
+	Checksum  uint32
+	Err       error
+}
+
+// StoreCheckpointV6 stores checkpoint file into a main file and 17 file parts.
+// the main file stores:
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+// 	the first 16 files parts contain the trie nodes below the subtrieLevel
+//	the last part file contains the top level trie nodes above the subtrieLevel and all the trie root nodes.
+func StoreCheckpointV6(
+	tries []*trie.MTrie, outputDir string, outputFile string, logger *zerolog.Logger) error {
+	if len(tries) == 0 {
+		logger.Info().Msgf("no tries to be checkpointed")
+		return nil
+	}
+
+	first, last := tries[0], tries[len(tries)-1]
+	logger.Info().
+		Str("first_hash", first.RootHash().String()).
+		Uint64("first_reg_count", first.AllocatedRegCount()).
+		Str("last", last.RootHash().String()).
+		Uint64("last_reg_count", last.AllocatedRegCount()).
+		Msgf("storing checkpoint for %v tries to %v", len(tries), outputDir)
+
+	subtrieRoots := createSubTrieRoots(tries)
+
+	estimatedSubtrieNodeCount := estimateSubtrieNodeCount(tries)
+
+	subTrieRootIndices, subTriesNodeCount, subTrieChecksums, err := storeSubTrieConcurrently(
+		subtrieRoots,
+		estimatedSubtrieNodeCount,
+		outputDir,
+		outputFile,
+		logger,
+	)
+	if err != nil {
+		return fmt.Errorf("could not store sub trie: %w", err)
+	}
+
+	logger.Info().Msgf("subtrie have been stored. sub trie node count: %v", subTriesNodeCount)
+
+	topTrieChecksum, err := storeTopLevelNodesAndTrieRoots(
+		tries, subTrieRootIndices, subTriesNodeCount, outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store top level tries: %w", err)
+	}
+
+	err = storeCheckpointSummary(subTrieChecksums, topTrieChecksum, outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store checkpoint summary: %w", err)
+	}
+
+	logger.Info().Msgf("checkpoint file has been successfully stored at %v/%v", outputDir, outputFile)
+
+	return nil
+}
+
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+func storeCheckpointSummary(
+	subTrieChecksums []uint32,
+	topTrieChecksum uint32,
+	outputDir string,
+	outputFile string,
+	logger *zerolog.Logger,
+) error {
+	closable, err := createWriterForCheckpointSummary(outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store checkpoint summary: %w", err)
+	}
+	defer func() {
+		closeErr := closable.Close()
+		// Return close error if there isn't any prior error to return.
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	// write version
+	version := encodeVersion(MagicBytes, VersionV6)
+	writer := NewCRC32Writer(closable)
+	_, err = writer.Write(version)
+	if err != nil {
+		return fmt.Errorf("cannot write checksum summary header: %w", err)
+	}
+
+	// TODO: write checksums
```

Comment:
I think you can remove this TODO

Reply:
TODOs are done and cleaned in a different PR